core/src/main/scala/kafka/admin/ZkSecurityMigrator.scala

```diff
@@ -0,0 +1,209 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kafka.admin
+
+import java.util.concurrent.atomic.AtomicInteger
+import java.util.concurrent.LinkedBlockingQueue
+import java.util.concurrent.ThreadPoolExecutor
+import java.util.concurrent.TimeUnit
+import javax.security.auth.login.LoginException
+import joptsimple.OptionParser
+import org.I0Itec.zkclient.ZkClient
+import org.I0Itec.zkclient.exception.ZkException
+import kafka.utils.{ToolsUtils, Logging, ZKGroupTopicDirs, ZkUtils, CommandLineUtils, ZkPath}
+import org.apache.kafka.common.security.JaasUtils
+
+import org.apache.zookeeper.AsyncCallback.{ChildrenCallback, StatCallback}
+import org.apache.zookeeper.CreateMode
+import org.apache.zookeeper.data.Stat
+import org.apache.zookeeper.KeeperException
+import org.apache.zookeeper.KeeperException.Code
+import org.apache.zookeeper.ZooDefs.Ids
+import org.apache.zookeeper.ZooKeeper
+
+import scala.collection.JavaConverters._
+import scala.collection._
+import scala.concurrent._
+
+object ZkSecurityMigrator extends Logging {
+  val counter = new AtomicInteger(0)
+  var exception: ZkException = null
+  val workQueue = new LinkedBlockingQueue[Runnable]
+  val threadExecutionContext = 
+    ExecutionContext.fromExecutor(new ThreadPoolExecutor(1,
+                                                         Runtime.getRuntime().availableProcessors(),
+                                                         5000,
+                                                         TimeUnit.MILLISECONDS,
+                                                         workQueue))
```

Comment:
I had a quick look into what you're saying, but opted for this way of building the executor because I have experience with it. If doing differently has any benefit, then it is not a problem to change.